src/vivarium_census_prl_synth_pop/data/generate_business_names.py

```diff
@@ -0,0 +1,148 @@
+        compression="bz2"
+    )
+
+
+def make_bigrams(df: pd.DataFrame) -> defaultdict:
+    """
+
+    Parameters
+    ----------
+    df: A one column dataframe from a CSV containing a list of real business names.
+
+    Makes default dict of business names for map to sample from
+    Bigrams will be a Dict[str: Dict[str: int]]
+
+    Example:  A series of business names to be turned into a Dict map
+            location_name
+        0 	Cottage Inn Pizza
+        1 	Union St Barber Shop
+        2 	Larkfield Auto Center
+        3 	Randall Guns
+        4 	Sport Clips
+
+    Create bigrams to map a word and the frequence at which the next choice happens.
+
+        {'<start>': {'Cottage': 1,
+                     'Union': 1,
+                     'Larkfield': 1,
+                     'Randall': 1,
+                     'Sport': 1
+                     }
+        }
+
+    Go through each word in each name and get the frequency of each name pair in the series
+    Returns
+    -------
+    defaultdict
+    """
+
+    def dict_factory():
+        return lambda: defaultdict(int)
+
+    bigrams = defaultdict(dict_factory())
+    n_rows = len(df)  # expect a few minutes
+
+    for i in range(n_rows):
+        if i % 10_000 == 0:
+            print(".", end=" ", flush=True)
+        names_i = df.iloc[i, 0]
+
+        tokens_i = names_i.split(" ")
+        for j in range(len(tokens_i)):
+            if j == 0:
+                bigrams["<start>"][tokens_i[j]] += 1
+            else:
+                bigrams[tokens_i[j - 1]][tokens_i[j]] += 1
+        bigrams[tokens_i[j]]["<end>"] += 1
+
+    return bigrams
+
+
+def sample_names(bigrams: defaultdict, n_businesses: int, n_max_tokens: int) -> pd.Series:
+    """
+
+    Parameters
+    ----------
+    bigrams: Default dict produced from make_bigrams function (see formatting of default dict.
+    n_businesses: Int of how many business names to generate
+    n_max_tokens: Int of max number of words possible for a business name to contain
+
+    Returns
+    -------
+    A string that is a randomly generated business name
+    """
+
+    logger.info("")
+    columns = [f"word_{i}" for i in range(n_max_tokens)]
+    names = pd.DataFrame(columns=columns)
+    names["word_0"] = ["<start>"] * n_businesses
+
+    for i in range(1, n_max_tokens):
+        if i == 1:
+            logger.info("Initializing random name sampling for business name generator")
+        previous_word = f"word_{i - 1}"
+        next_word = f"word_{i}"
+        current_words_count_dict = names[previous_word].value_counts().to_dict()
+        for word in current_words_count_dict.keys():
+            logger.info(
+                f"Generating {len(list(current_words_count_dict.keys()))} words "
+                f"in {i}th of {n_max_tokens} columns for business names data."
+            )
+            if word != "<end>":
+                vals = list(bigrams[word].keys())
+                pr = np.array(list(bigrams[word].values()))
```

Comment:
maybe name this something else? because right now `np.array(list(bigrams[word].values()))` is a frequency, and then it becomes a probability on line 132 when you do `pr / pr.sum()`